core/src/main/java/org/elasticsearch/common/xcontent/ObjectParser.java

```diff
@@ -89,6 +91,12 @@
     private final boolean ignoreUnknownFields;
 
     /**
+     * A special purpose field parser that gets used when the provided matchFieldPrecidate accepts it
+     */
+    private FieldParser matchFieldParser;
+    private Predicate<String> matchFieldPredicate;
```

Comment:
I can combine these into an inner private struct-like class. I think they should be separate arguments in the `declareMatchFieldParser` method though, that way it's easier to use and we don't need to expose another class for this.